compiler/src/iree/compiler/Codegen/Common/ConvertBf16ArithToF32.cpp

```diff
@@ -0,0 +1,312 @@
+    }
+
+    Operation *newOp = rewriter.create(state);
+    rewriter.replaceOp(op, newOp->getResults());
+    return success();
+  }
+};
+
+template <typename OpTy, typename TypeTy,
+          typename OperandToResultWidthLegalityRelation>
+struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, typename OpTy::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto resultType =
+        this->getTypeConverter()->convertType(op.getResult().getType());
+    auto operandType =
+        this->getTypeConverter()->convertType(op.getOperand().getType());
+
+    auto resultEType = cast<TypeTy>(getElementTypeOrSelf(resultType));
+    auto operandEType = cast<TypeTy>(getElementTypeOrSelf(operandType));
+    // If post-conversion, the types would be equal, then the op becomes a
+    // no-op. Note that the op does not itself allow such a configuration, so we
+    // have to catch this before creating the new op.
+    if (resultEType == operandEType) {
+      rewriter.replaceOp(op, adaptor.getOperands()[0]);
+      return success();
+    }
+    // If after conversion the op becomes invalid, but not same-type (which we
+    // can fold above), then bail out.
+    // TODO: In some cases, we can repair the situation here, but for integer
+    // truncation, we don't know whether we should invert with signed or
+    // unsigned extension.
+    if (!OperandToResultWidthLegalityRelation()(operandEType.getWidth(),
+                                                resultEType.getWidth())) {
+      return rewriter.notifyMatchFailure(op, "invalid width combination");
+    }
+    rewriter.replaceOpWithNewOp<OpTy>(op, resultType, op.getOperand());
+    return success();
+  }
+};
+
+template <typename SrcOp>
+class PropagateCastF : public OpRewritePattern<SrcOp> {
+  using OpRewritePattern<SrcOp>::OpRewritePattern;
+
+ public:
+  LogicalResult matchAndRewrite(SrcOp op,
+                                PatternRewriter &rewriter) const override {
+    auto operand = op.getOperand();
+    auto ty = dyn_cast<VectorType>(operand.getType());
+    auto resultTy = dyn_cast<VectorType>(op.getType());
+
+    if (!ty || ty.getRank() != 0) {
+      return rewriter.notifyMatchFailure(op, "Not casting from vector-scalar");
+    }
+
+    mlir::DenseElementsAttr vectorCst;
+    if (!matchPattern(operand, m_Constant(&vectorCst))) {
+      return failure();
+    }
+
+    mlir::FloatAttr val = vectorCst.getSplatValue<mlir::FloatAttr>();
+    auto newVal = FloatAttr::get(resultTy.getElementType(),
+                                 val.getValue().convertToDouble());
+    auto vectorVal = DenseElementsAttr::get(resultTy, newVal);
+    rewriter.replaceOpWithNewOp<arith::ConstantOp>(op, resultTy, vectorVal);
+    return success();
+  }
+};
+
+template <typename Base, typename Converter>
```

Comment:
Do we really need this to be template parameters? There is only one instantiation of this.

Reply:
It does not, it was inherited from the forked passes.. Removed the templated behavior.